Automattic/woocommerce-payments
# Add support for allowedShippingCountries in Express Checkout Element

Implements https://github.com/Automattic/woocommerce-payments/issues/9104

## Changes proposed in this Pull Request

Adds support for specifying allowed shipping countries (`allowedShippingCountries`) in the Express Checkout Element.

## Testing instructions

> [!TIP]  
> If you want to test with multiple addresses from different countries, you can add yourself to the Google test card suite group. In test mode, you’ll have access to a set of test cards, rather than the cards associated with your production account.
 https://developers.google.com/pay/api/web/guides/resources/test-card-suite

> [!NOTE]  
> To test the changes introduced here, we first need to create shipping options for at least two countries. In this case the countries will be USA and Brazil. You can create a "Flat rate" shipping method for both countries. That will be sufficient.

> [!NOTE]  
> The test results should be the same for all the pages affected by this PR: Product Page, Cart Page, Checkout Page, Cart shortcode, and Checkout shortcode. (Does not apply to the "Pay for Order" page.)

> [!NOTE]
> All cases should be tested in the same way:
>1. If you are on the page of a physical product (that is, it requires shipping), simply click the Express Checkout button and proceed to checkout.
>2. If you are testing on other pages (except "Pay for Order"), add a product to the cart and then visit the desired page. Click the Express Checkout button and proceed to checkout.
>3. **There's no need to test the "Pay for Order"** page since all orders on this page are defined [as not requiring shipping](https://github.com/Automattic/woocommerce-payments/blob/7f8cad70de8488f5bdac3599d21697b50b1e2416/client/express-checkout/index.js#L594) in the Express Checkout settings.

> [!NOTE]
> Since this PR includes many combinations to test, and given that ECE has two different implementations, I would suggest testing at least the Checkout (block) page and the Shortcode Checkout page. **That would cover both ECE implementations currently in WooPayments**.

> [!IMPORTANT]
> Test with Apple Pay on at least the pages mentioned above. The results should be the same for each test.

### Case 1: "Shipping location(s)" is set to "Ship to all countries you sell to"

- **"Selling location(s)" is set to "Sell to all countries"**
  - ✅ The checkout process should work without issues for any shipping address from any country.
  - Note that it won't be possible to checkout using a shipping address from a country without shipping options.
- **"Selling location(s)" is set to "Sell to all countries, except for…"** and "Sell to all countries, except for…" is set to "United Kingdom (UK)"
  - ✅ Checkout should work without issues for any shipping address except one from the UK.
  - Note that it won’t be possible to checkout using a shipping address from a country without shipping options. For this PR, it should be possible to checkout with shipping addresses from the USA and Brazil only, as these are the only countries with shipping options.
- **"Selling location(s)" is set to "Sell to specific countries"** and “Sell to specific countries” is set to "United States (US)"
  - ✅ Checkout should complete without issues with any US address.
  - It shouldn't be possible to select a shipping address from Brazil or any country other than the USA.

### Case 2: "Shipping location(s)" is set to "Ship to all countries"

This option ignores the value set in "Selling location(s)", therefore:

- ✅ Checkout should complete without issues for all countries.
- Note that it won't be possible to checkout using a shipping address from a country without shipping options.
- Therefore, for this case, it should allow shipping to the USA and Brazil. If desired, you can add another country with shipping options to verify that shipping works to that country as well.

### Case 3: "Shipping location(s)" is set to "Ship to specific countries only" and "Ship to specific countries" is set to "Brazil"

- ✅ Checkout should complete without issues with shipping addresses from Brazil.
- It shouldn’t be possible to select a shipping address from any other country.

### Potential Issues:

If, during testing, you find that you can’t open the payment sheet and see the following error in the browser console:


```
Uncaught (in promise) IntegrationError: When `shippingAddressRequired` is true, you must specify `shippingRates`.
```


This is a known issue and has been reported [here](https://github.com/Automattic/woocommerce-payments/issues/9593). One way to fix it is by clearing all cookies and starting again. Another way is to go to the checkout page and change the shipping address to a country with shipping options, such as the USA or Brazil.

-------------------


**Post merge**
